internal/provision/profiles.go

```diff
@@ -34,81 +37,156 @@ const (
+		secretProvider := bootstrapContainer.SecretProviderFrom(dic.Get)
+		addProfilesReq, edgexErr = loadProfilesFromURI(path, parsedUrl, dpc, secretProvider, lc)
+		if edgexErr != nil {
+			return edgexErr
+		}
+	} else {
+		addProfilesReq, edgexErr = loadProfilesFromFile(path, dpc, lc)
+		if edgexErr != nil {
+			return edgexErr
+		}
+	}
+
+	if len(addProfilesReq) == 0 {
+		return nil
+	}
+	ctx := context.WithValue(context.Background(), common.CorrelationHeader, uuid.NewString()) // nolint:staticcheck
+	_, edgexErr = dpc.Add(ctx, addProfilesReq)
+	return edgexErr
+}
+
+func loadProfilesFromFile(path string, dpc interfaces.DeviceProfileClient, lc logger.LoggingClient) ([]requests.DeviceProfileRequest, errors.EdgeX) {
+	var edgexErr errors.EdgeX
 	absPath, err := filepath.Abs(path)
 	if err != nil {
-		return errors.NewCommonEdgeX(errors.KindServerError, "failed to create absolute path", err)
+		return nil, errors.NewCommonEdgeX(errors.KindServerError, "failed to create absolute path for profiles", err)
 	}
 
 	files, err := os.ReadDir(absPath)
 	if err != nil {
-		return errors.NewCommonEdgeX(errors.KindServerError, "failed to read directory", err)
+		return nil, errors.NewCommonEdgeX(errors.KindServerError, "failed to read directory for profiles", err)
 	}
 
 	if len(files) == 0 {
-		return nil
+		return nil, nil
 	}
 
-	lc := bootstrapContainer.LoggingClientFrom(dic.Get)
 	lc.Infof("Loading pre-defined profiles from %s(%d files found)", absPath, len(files))
-
-	var addProfilesReq []requests.DeviceProfileRequest
-	dpc := bootstrapContainer.DeviceProfileClientFrom(dic.Get)
+	var addProfilesReq, processedProfilesReq []requests.DeviceProfileRequest
 	for _, file := range files {
-		var profile dtos.DeviceProfile
 		fullPath := filepath.Join(absPath, file.Name())
-		if strings.HasSuffix(fullPath, yamlExt) || strings.HasSuffix(fullPath, ymlExt) {
-			content, err := os.ReadFile(fullPath)
-			if err != nil {
-				lc.Errorf("Failed to read %s: %v", fullPath, err)
-				continue
-			}
+		processedProfilesReq, edgexErr = processProfiles(fullPath, fullPath, nil, lc, dpc)
+		if edgexErr != nil {
+			lc.Error(edgexErr.Error())
+			return addProfilesReq, nil
+		}
+		if len(processedProfilesReq) > 0 {
+			addProfilesReq = append(addProfilesReq, processedProfilesReq...)
+		}
+	}
+	return addProfilesReq, nil
+}
 
-			err = yaml.Unmarshal(content, &profile)
-			if err != nil {
-				lc.Errorf("Failed to YAML decode profile %s: %v", file.Name(), err)
-				continue
-			}
-		} else if strings.HasSuffix(fullPath, jsonExt) {
-			content, err := os.ReadFile(fullPath)
-			if err != nil {
-				lc.Errorf("Failed to read %s: %v", fullPath, err)
-				continue
-			}
+func loadProfilesFromURI(inputURI string, parsedURI *url.URL, dpc interfaces.DeviceProfileClient, secretProvider bootstrapInterfaces.SecretProvider, lc logger.LoggingClient) ([]requests.DeviceProfileRequest, errors.EdgeX) {
+	var edgexErr errors.EdgeX
+	bytes, err := file.Load(inputURI, secretProvider, lc)
```

Comment:
Add comment that the is expected to be the index file containing list of profiles to load.